File: ios/LumaNox/Core/Telemetry/FirebaseTelemetry.swift
import FirebaseAnalytics
import FirebaseCore
import FirebaseCrashlytics
import Foundation
import os

enum FirebaseTelemetry {
    private static let log = Logger(subsystem: "com.xpx.vault", category: "FirebaseTelemetry")
    private(set) static var isConfigured = false

    static func configure() {
        guard !isConfigured else { return }

        if FirebaseApp.app() == nil {
            guard Bundle.main.path(forResource: "GoogleService-Info", ofType: "plist") != nil else {
                log.info("Firebase skipped because GoogleService-Info.plist is not bundled.")
                return
            }
            FirebaseApp.configure()
        }

        Analytics.setAnalyticsCollectionEnabled(!isDebugBuild)
        Crashlytics.crashlytics().setCrashlyticsCollectionEnabled(!isDebugBuild)
        isConfigured = true
    }

    static func logEvent(_ name: String, parameters: [String: String] = [:]) {
        guard isConfigured else { return }
        let payload = parameters.reduce(into: [String: Any]()) { result, item in
            result[sanitizeName(item.key)] = sanitizeValue(item.value)
        }
        Analytics.logEvent(sanitizeName(name), parameters: payload.isEmpty ? nil : payload)
    }

    static func breadcrumb(_ tag: String, _ message: String) {
        guard isConfigured else { return }
        Crashlytics.crashlytics().log("[\(sanitizeName(tag))] \(sanitizeValue(message))")
    }

    private static var isDebugBuild: Bool {
        #if DEBUG
        true
        #else
        false
        #endif
    }

    private static func sanitizeName(_ value: String) -> String {
        let filtered = value.filter { $0.isLetter || $0.isNumber || $0 == "_" }
        return String((filtered.isEmpty ? "event" : filtered).prefix(40))
    }

    private static func sanitizeValue(_ value: String) -> String {
        String(value.prefix(100))
    }
}

File: android/app/src/main/kotlin/com/xpx/vault/telemetry/FirebaseTelemetry.kt
package com.xpx.vault.telemetry

import android.app.Application
import android.os.Bundle
import com.google.firebase.FirebaseApp
import com.google.firebase.analytics.FirebaseAnalytics
import com.google.firebase.crashlytics.FirebaseCrashlytics
import com.xpx.vault.BuildConfig

/**
 * Firebase Analytics / Crashlytics adapter.
 *
 * The SDK is activated only when a valid google-services.json is present and
 * the Google Services Gradle plugin generated a default Firebase app.
 */
object FirebaseTelemetry {
    @Volatile
    private var analytics: FirebaseAnalytics? = null

    @Volatile
    private var crashlytics: FirebaseCrashlytics? = null

    fun install(application: Application) {
        if (FirebaseApp.getApps(application).isEmpty()) {
            return
        }
        analytics = FirebaseAnalytics.getInstance(application).apply {
            setAnalyticsCollectionEnabled(!BuildConfig.DEBUG)
        }
        crashlytics = FirebaseCrashlytics.getInstance().apply {
            setCrashlyticsCollectionEnabled(!BuildConfig.DEBUG)
        }
    }

    fun logEvent(name: String, params: Map<String, String>) {
        val instance = analytics ?: return
        instance.logEvent(sanitizeName(name), Bundle().apply {
            params.forEach { (key, value) ->
                putString(sanitizeName(key), sanitizeValue(value))
            }
        })
    }

    fun breadcrumb(tag: String, message: String) {
        crashlytics?.log("[${sanitizeName(tag)}] ${sanitizeValue(message)}")
    }

    private fun sanitizeName(value: String): String =
        value.filter { it.isLetterOrDigit() || it == '_' }
            .ifBlank { "event" }
            .take(MAX_NAME_LENGTH)

    private fun sanitizeValue(value: String): String = value.take(MAX_VALUE_LENGTH)

    private const val MAX_NAME_LENGTH = 40
    private const val MAX_VALUE_LENGTH = 100
}
